crates/astria-core/src/sequencerblock/v1/block/mod.rs

```diff
@@ -1165,122 +1162,424 @@ impl ExtendedCommitInfoError {
+    #[error(
+        "failed constructing the extended commit info with currency pair mapping from the raw \
+         protobuf"
+    )]
     InvalidExtendedCommitInfo(ExtendedCommitInfoWithCurrencyPairMappingError),
 }
 
-fn get_and_verify_extended_commit_info_proof(
-    extended_commit_info: Option<Bytes>,
-    extended_commit_info_proof: Option<crate::generated::primitive::v1::Proof>,
-    data_hash: [u8; 32],
-) -> Result<Option<merkle::Proof>, ExtendedCommitInfoError> {
-    use sha2::Digest as _;
+/// Wraps the initial elements of the `txs` field of CometBFT requests/responses like
+/// `PrepareProposal` and `FinalizeBlock`.
+///
+/// These special elements always appear before all the actual rollup txs in a defined order.
+///
+/// Rollup txs are not included in this enum as wrapping them would cause them to be indexed under
+/// a different hash from their original one by CometBFT, making it difficult for clients to
+/// identify their txs and meaning they would not be cleared out of the CometBFT mempool when added
+/// to a block.
+#[derive(BorshSerialize, BorshDeserialize)]
+#[expect(clippy::doc_markdown, reason = "false positive")]
+pub enum DataItem {
+    RollupTransactionsRoot([u8; SHA256_DIGEST_LENGTH]),
+    RollupIdsRoot([u8; SHA256_DIGEST_LENGTH]),
+    UpgradeChangeHashes(Vec<ChangeHash>),
+    ExtendedCommitInfo(Bytes),
+}
 
-    let extended_commit_info_proof = if let Some(extended_commit_info) = extended_commit_info {
-        let Some(extended_commit_info_proof) = extended_commit_info_proof else {
-            return Err(ExtendedCommitInfoError::proof_not_set());
-        };
-        let extended_commit_info_proof = merkle::Proof::try_from_raw(extended_commit_info_proof)
-            .map_err(ExtendedCommitInfoError::invalid_extended_commit_info_proof)?;
+impl DataItem {
+    /// The number of bytes output by [`Self::encode`] when `self` is `RollupIdsRoot`.
+    pub const ENCODED_ROLLUP_IDS_ROOT_LENGTH: usize = SHA256_DIGEST_LENGTH + 1;
+    /// The number of bytes output by [`Self::encode`] when `self` is `RollupTransactionsRoot`.
+    pub const ENCODED_ROLLUP_TRANSACTIONS_ROOT_LENGTH: usize = SHA256_DIGEST_LENGTH + 1;
+    const EXTENDED_COMMIT_INFO_VARIANT_NAME: &'static str = "extended commit info";
+    const ROLLUP_IDS_ROOT_VARIANT_NAME: &'static str = "rollup ids root";
+    const ROLLUP_TXS_ROOT_VARIANT_NAME: &'static str = "rollup transactions root";
+    const UPGRADE_HASHES_VARIANT_NAME: &'static str = "upgrade change hashes";
+
+    fn variant_name(&self) -> &'static str {
+        match self {
+            DataItem::RollupTransactionsRoot(_) => Self::ROLLUP_TXS_ROOT_VARIANT_NAME,
+            DataItem::RollupIdsRoot(_) => Self::ROLLUP_IDS_ROOT_VARIANT_NAME,
+            DataItem::UpgradeChangeHashes(_) => Self::UPGRADE_HASHES_VARIANT_NAME,
+            DataItem::ExtendedCommitInfo(_) => Self::EXTENDED_COMMIT_INFO_VARIANT_NAME,
+        }
+    }
+
+    /// Returns the Borsh-encoded bytes of `self`.
```

Comment:
once again, why borsh here? very surprising

Reply:
Agree it's surprising, but as before, we need deterministic encoding.

Reply:
i don't understand why `DataItem` encoding needs to be deterministic while user txs can be protobuf - the encoding of all transactions is canonicalized in the block data itself. this is putting the `DataItem`s as borsh-encodded into the block itself right?

Reply:
> We need deterministic encoding to guarantee identical hashes across validators.

this is only for upgrade hashes within the DataItem though right?

Reply:
> this is only for upgrade hashes within the DataItem though right?

True.  In this case I was thinking we'd want all nodes to generate identical encoded output so that the byte lengths were guaranteed identical, so we can reach consensus around block size limits.  Maybe we can get that using protobuf-encoding, but since these encoded data items aren't part of our API (or at least I don't consider them part of our API - maybe I'm wrong?) I thought it was better to just use a schema with guaranteed outputs, and avoid polluting our protobuf API with a new `DataItem` message type which is really just an internal implementation detail.

Reply:
> we'd want all nodes to generate identical encoded output 
there's only one node that needs to generate this output though, which is the proposer, the other nodes would just decode and check the data. 

>  avoid polluting our protobuf API with a new DataItem message type which is really just an internal implementation detail. 
I somewhat agree, but the fact that the DataItems end up in a block makes them "public"/non-internal in a way. I think it's fine to add these as proto definitions, then all transactions whether injected or user-submitted can be proto encoded

Reply:
Changed to protobuf-encoding in [ad1a8ff](https://github.com/astriaorg/astria/pull/1917/commits/ad1a8ffb511268112b522fc61d045a3ad4c11a3e).